aws/logs_monitoring/cache.py

```diff
@@ -0,0 +1,421 @@
+            )
+            additional_tags = [
+                f"http_status_code:{e.response['ResponseMetadata']['HTTPStatusCode']}"
+            ]
+            send_forwarder_internal_metrics(
+                "client_error", additional_tags=additional_tags
+            )
+
+        logger.debug(
+            "Built this tags cache from GetResources API calls: %s", tags_by_arn_cache
+        )
+
+        return tags_fetch_success, tags_by_arn_cache
+
+    def get(self, key):
+        """Get the tags for the Lambda function from the cache
+
+        Will refetch the tags if they are out of date, or a lambda arn is encountered
+        which isn't in the tag list
+
+        Note: the ARNs in the cache have been lowercased, so resource_arn must be lowercased
+
+        Args:
+            key (str): the key we're getting tags from the cache for
+
+        Returns:
+            lambda_tags (str[]): the list of "key:value" Datadog tag strings
+        """
+        if self._is_expired():
+            send_forwarder_internal_metrics("local_cache_expired")
+            logger.debug("Local cache expired, fetching cache from S3")
+            self._refresh()
+
+        function_tags = self.tags_by_id.get(key, [])
+        return function_tags
+
+
+#############################
+# Cloudwatch Log Group Tags #
+#############################
+
+cloudwatch_logs_client = boto3.client("logs")
+
+
+def get_log_group_tags(log_group):
+    response = None
+    try:
+        send_forwarder_internal_metrics("list_tags_log_group_api_call")
+        response = cloudwatch_logs_client.list_tags_log_group(logGroupName=log_group)
+    except Exception as e:
+        logger.exception(f"Failed to get log group tags due to {e}")
+    if response is not None:
+        formatted_tags = [
+            "{key}:{value}".format(
+                key=sanitize_aws_tag_string(k, remove_colons=True),
+                value=sanitize_aws_tag_string(v, remove_leading_digits=False),
+            )
+            if v
+            else sanitize_aws_tag_string(k, remove_colons=True)
+            for k, v in response["tags"].items()
+        ]
+    else:
+        # In the case of failing to query tags we will write an empty list and retry on the next refresh
+        formatted_tags = []
+    return formatted_tags
+
+
+class CloudwatchLogGroupTagsCache(LambdaTagsCache):
+    CACHE_FILENAME = DD_S3_LOG_GROUP_CACHE_FILENAME
+    CACHE_LOCK_FILENAME = DD_S3_LOG_GROUP_CACHE_LOCK_FILENAME
+
+    def build_tags_cache(self):
+        """Makes API calls to GetResources to get the live tags of the account's Lambda functions
+
+        Returns an empty dict instead of fetching custom tags if the tag fetch env variable is not set to true
+
+        Returns:
+            tags_by_arn_cache (dict<str, str[]>): each Lambda's tags in a dict keyed by ARN
+        """
+        new_tags = {}
+        for log_group in self.tags_by_id.keys():
+            new_tags[log_group] = get_log_group_tags(log_group)
```

Comment:
Here we're making multiple api calls to refill the cache, so I don't want to return false and not write to the cache if one of those api calls fails. I suppose if all the api calls fail we could return False here and skip writing to the s3 cache.
If the api call fails we can take the value we have stored from the local cache to avoid overwriting on a failed api call.